### ISSUE TITLE
`tf.data` shuffle uses 2.5x more memory than necessary during `model.fit`

### ISSUE THREAD

**Issue by lgeiger** [tensorflow/tensorflow#54299]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.8.0 and 2.9.0.dev20220208
- Python version: 3.9

**Describe the current behavior**

When training a Keras model with a `tf.data.Dataset` that has not been cached in memory before applying `tf.data.Dataset.shuffle(buffer)`, it seems like the shuffle buffer is never properly freed thus memory usage increases far beyond the memory expected from the shuffle buffer.

When running the code below using `dataset.cache().shuffle(num_examples)` the entire dataset will be loaded into memory and shuffled. This is expected and the code will require **~145 GB** of memory which is equivalent to the size of the dataset.

However using `dataset.shuffle(num_examples // 2)` without prior in memory caching (like what would be require on smaller machines) the code requires **~160 GB** of memory which is more than the entire size of the dataset, making `.shuffle()` unusable with large datasets and buffers. The same behaviour can be observed when setting `reshuffle_each_iteration=False`. It seems like after the first epoch the memory usage just continues to go up rather than staying at roughly the size that is required to store the shuffle buffer.

**Describe the expected behavior**
I would expect that `tf.data` and `model.fit` do not use memory beyond what's set required by the shuffle buffer, so in this example around **~73 GB**. Ideally the buffer would even be shared across epochs so that it doesn't need to be filled again before every epoch. A bit of memory overhead is expected to do this efficiently but the overhead should not be twice the size of the shuffle buffer and should certainly not exceed the total size of the dataset since that defeats the purpose of using a shuffle buffer that is smaller than the dataset.

I don't think I am doing anything non-standard in the code below. In fact I follow strickly the recommended usage of both `tf.data` and Keras so I am very surprised about the larger than expected memory usage.

**Standalone code to reproduce the issue**

Consider the following simple example that trains a no-op model on a large dataset (in this case ImageNet):
```python
import tensorflow as tf
import tensorflow_datasets as tfds
from tensorflow import keras

batch_size = 1024

dataset, info = tfds.load(
    "imagenet2012:5.0.*",
    decoders={"image": tfds.decode.SkipDecoding()},
    split="train",
    with_info=True,
    data_dir="gs://my_data_bucket",
)
num_examples = info.splits["train"].num_examples


def _decode_and_center_crop(image_bytes):
    """Crops to center of image with padding then scales image_size."""
    shape = tf.image.extract_jpeg_shape(image_bytes)
    image_height, image_width = shape[0], shape[1]
    image_size = 224

    padded_center_crop_size = tf.cast(
        (
            (image_size / (image_size + 32))
            * tf.cast(tf.minimum(image_height, image_width), tf.float32)
        ),
        tf.int32,
    )

    offset_height = ((image_height - padded_center_crop_size) + 1) // 2
    offset_width = ((image_width - padded_center_crop_size) + 1) // 2
    crop_window = tf.stack(
        [offset_height, offset_width, padded_center_crop_size, padded_center_crop_size]
    )
    image = tf.image.decode_and_crop_jpeg(image_bytes, crop_window, channels=3)
    image = tf.image.resize(image, [image_size, image_size])
    return tf.cast(image, dtype=tf.float32)


def preprocessing(data):
    return _decode_and_center_crop(data["image"]), data["label"]


dataset = (
    dataset.shuffle(num_examples // 2)
    .map(preprocessing, num_parallel_calls=tf.data.AUTOTUNE)
    .batch(batch_size)
    .prefetch(1)
)

model = keras.models.Sequential(
    [
        keras.layers.GlobalMaxPool2D(input_shape=(224, 224, 3)),
        keras.layers.Dense(1000, activation="softmax"),
    ]
)

model.compile(optimizer="sgd", loss="sparse_categorical_crossentropy")

model.fit(dataset, epochs=3)
```

Unfortunately this cannot be reproduced on Colab since not enough memory is available there. This issue is not related the the exact dataset used. I am just using ImageNet streamed from a GCS bucket, since this should be a pretty standard task. However this can be reproduced with any significantly large dataset.

The underlying cause might similar to what caused #36240, but I haven't been able to look into the `tf.data` or Keras code to debug what's going on.

**Comment by Wansh619** [comment]
Hi @lgeiger i am ne in this open source stuff and want to learn to work on this issues so can you assign me for this issue and provide me resources to learn for solving this issue?

**Comment by lgeiger** [comment]
> Hi @lgeiger i am ne in this open source stuff and want to learn to work on this issues so can you assign me for this issue and provide me resources to learn for solving this issue?

Unfortunately I can't. I am not familiar with the `tf.data` code myself and don't have the required repo access either.

**Comment by Wansh619** [comment]
I am a absolute beginner so i have to learn from start  can you just tell me form where i can learn all these things?

**Comment by Wansh619** [comment]
Sorry but i want to know about that what to learn for solving this issue do i have learn about somthing specifice in tensorflow lib or anything else .

I am asking for material to work on source code not for how to contribute

**Comment by aaudiber** [comment]
@lgeiger Could the issue be reproduced in colab by using a smaller shuffle buffer? Also, how are you measuring memory usage?

**Comment by lgeiger** [comment]
> @lgeiger Could the issue be reproduced in colab by using a smaller shuffle buffer?

I haven't tried since I am not sure how one would properly measure the memory usage within Colab. But in general it might be a bit tricky to reproduce the issue with a smaller shuffle buffer as the memory increase could be hidden by the normal memory usage of TensorFlow which tends to fluctuate a bit.

> Also, how are you measuring memory usage?

I ran the code on a fresh GCP VM with lots of memory and visually look at the memory usage reported by [`bottom`](https://github.com/ClementTsang/bottom) (but any other system monitor should do).

**Comment by lgeiger** [comment]
@aaudiber were you able to reproduce this issue on your end?
I am still seeing the same behaviour in TF 2.9.1 which makes proper shuffling of datasets that do not fit into memory pretty much impossible together with Keras `model.fit`.
Let me know if I can provide you with more information about this problem.

**Comment by aaudiber** [comment]
@lgeiger Is it possible to reproduce the issue without `model.fit` by iterating through the dataset directly? That could help narrow down whether the issue is in `shuffle` itself as opposed to the way that the dataset is used by `model.fit`.

Also, when measuring memory usage make sure you aren't counting the operating system's buffer cache. You can free the buffer cache before measuring memory with the commands described in: https://www.baeldung.com/linux/empty-buffer-cache#1-proc-sys-vm-drop-caches-command

**Comment by lgeiger** [comment]
@aaudiber To check whether this could somehow be a problem with Python's GC not cleaning up `data_iterator`, I also tested
```python
for epoch in range(3):
    for batch in iter(dataset):
        pass
```
and
```python
for epoch in range(3):
    for batch in dataset:
        pass
```
but both snippets show the same increasing memory usage as the original reproduction (tested on `f-nightly==2.10.0.dev20220711`).

**Comment by lgeiger** [comment]
Thanks for the minimal example and trying to reproduce the issue on your end!

I ran the above code and recorded memory usage with a [simple memory profiler](https://github.com/pythonprofilers/memory_profiler). I am running on a 48 core GCP node with 256 GB of memory and a single NVIDIA T4 attached. Softwarewise I just tried one of the default ML images `c1-deeplearning-tf-2-9-cu113-v20220701-debian-10` and installed `tf-nightly==2.10.0.dev20220711` from pip.

Unfortunately, the code still leaks memory:
![mem_1](https://user-images.githubusercontent.com/13285808/178597460-8164dfa3-aef5-4391-9a6e-85cde635836d.png)
![mem_half](https://user-images.githubusercontent.com/13285808/178597462-73e0a871-a4f1-453e-8535-d0452d6c3f55.png)
![mem_full](https://user-images.githubusercontent.com/13285808/178598146-fd2acd06-456a-4b91-bfc5-e6929147908a.png)

> Random guess, but perhaps is the difference in behavior is related to malloc memory fragmentation? My testing uses tcmalloc, which is better at avoiding fragmentation than the default malloc: https://github.com/google/tcmalloc

Thanks for the pointer. I will take a look it tcmalloc tomorrow.

**Comment by lgeiger** [comment]
@aaudiber Thanks again for your help. I tested out using tcmalloc using:
```
sudo apt install libtcmalloc-minimal4

export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4
```

And memory usage is completely stable across all shuffle buffer sizes:
![mem_1_tcmalloc](https://user-images.githubusercontent.com/13285808/178718302-8baa9781-b792-4f25-8532-2f709ab3f658.png)
![mem_half_tcmalloc](https://user-images.githubusercontent.com/13285808/178718307-b3586a31-f71d-4705-8558-58f518822898.png)
![mem_full_tcmalloc](https://user-images.githubusercontent.com/13285808/178718311-628bb00b-6e4e-45b4-9405-821485b663ae.png)


I am not sure whether using `LD_PRELOAD` has any downsides, but I guess the only alternative would be to build TensorFlow from source.
Considering the massive difference in memory usage, is the TensorFlow team considering building OSS TensorFlow against tcmalloc by default?

**Comment by Collonville** [comment]
> @aaudiber Thanks again for your help. I tested out using tcmalloc using:
> 
> ```
> sudo apt install libtcmalloc-minimal4
> 
> export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4
> ```
> 
> And memory usage is completely stable across all shuffle buffer sizes: ![mem_1_tcmalloc](https://user-images.githubusercontent.com/13285808/178718302-8baa9781-b792-4f25-8532-2f709ab3f658.png) ![mem_half_tcmalloc](https://user-images.githubusercontent.com/13285808/178718307-b3586a31-f71d-4705-8558-58f518822898.png) ![mem_full_tcmalloc](https://user-images.githubusercontent.com/13285808/178718311-628bb00b-6e4e-45b4-9405-821485b663ae.png)
> 
> I am not sure whether using `LD_PRELOAD` has any downsides, but I guess the only alternative would be to build TensorFlow from source. Considering the massive difference in memory usage, is the TensorFlow team considering building OSS TensorFlow against tcmalloc by default?

I had the same issue in TF2.9.1 but it worked well for this solution. Thanks for the research!

**Comment by ghylander** [comment]
this is a great find!
thanks for the extensive research and findings, the tcmmalloc and LD_PREOAD trick was great

Let's hope this gets 'fixed' at some point

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
